pkg/param/pagination/pagination.go

```diff
@@ -1,30 +1,33 @@
 package pagination
 
+// GetPage - pagination
 type GetPage struct {
 	StartRow  *int
 	EndRow    *int
 	PageSize  *int
```

Comment:
That already exist, you can set StartRow and EndRow .

Reply:
Why do we want multiple ways to specify the request parameters?  The meaning isn't clear if all the values are specified.

From the test case, the expected flow appears to be:

    arg.Page = new(pagination.Page)
    arg.Page.SetPageSize(2)
    volumes, err := suite.volumeService.GetVolumes(arg)
    for arg.Page.ContainsMore() {
            volumes, err := suite.volumeService.GetVolumes(arg)
    }

I find it unexpected that the GetVolumes function modifies the arg value.  We could clarify it by explicitly returning the Page values from the response.  I suggest a flow like:

	arg.Page = new(pagination.Page)
	var retrievedPage pagination.Page
	for hasNextPage := true; hasNextPage; hasNextPage = retrievedPage.HasNextPage() {
		arg.Page.NextPage(retrievedPage, 2)
		volumes, retrievedPage, err := suite.volumeService.GetVolumes(arg)
	}

The supporting function could look something like:

    func (page *Page) NextPage(retrievedPage Page, pageSize int) {
    	if retrievedPage == nil || retrievedPage.EndRow == nil {
    		page.SetStartRow(0)
    	} else {
    		page.SetStartRow(*retrievedPage.EndRow)
    	}
    	page.SetEndRow(*page.StartRow + pageSize)
    }

Reply:
>>Why do we want multiple ways to specify the request parameters? The meaning isn't clear if all the values are specified.
We are exposing only methods to set the page size , start row and end row for end user .  We arent exposing convenience method for TotalRow as it is not expected to be set by user . 

StartRow and EndRow add more flexibility to sdk user that's all. If we don't want we can remove it . 

`arg.Page = new(pagination.Page)
var retrievedPage pagination.Page
for hasNextPage := true; hasNextPage; hasNextPage = retrievedPage.HasNextPage() {
	arg.Page.NextPage(retrievedPage, 2)
	volumes, retrievedPage, err := suite.volumeService.GetVolumes(arg)
}`

This code will not work. As retrievedPage is a new variable. HasNextPage() will be false all the time. 

What is the use of this arg.Page.NextPage(retrievedPage, 2) , this is unnecessary . Intentions of having ContainsMore() is to jsut fetch the next Page so that user easily iterative over it.

Reply:
I should have been more clear.  In group_mgmt_client.go, the end of the ListFromParams function would look something like:

	returnedPage := pagination.Page { &wrapper.StartRow, &wrapper.EndRow, &wrapper.TotalRows }
	return wrapper.Data, returnedPage, nil

It is a question of who has control over which page is retrieved.  I feel the "advance to next page" logic should not be hidden inside the ListFromParams function.  Rather, the caller should explicitly indicate the desired page parameters.  The Page.NextPage function provides a convenience to set the StartRow and EndRow based on the previously retrieved page and the desired page size.  Note that the suggested client code is just as compact, but it gives the client more control and improved clarity.

Reply:
We dont want to give the user any selective page control as this may not be useful and he may not need it. TotalRow is internal to SDK as it is based on REST API return. 
When you make a call to REST API, it returns TotalRows . 
Let's consider an example: let's say there are 100s of volume on the array and the user wants to process all of them so instead of fetching all the volumes at one time, the user would set pageSize = 10, he would iterate over 10 volumes at a time and process. 
Tell me the use case where the user would like to retrieve a selective page? 

Here pagination is not used for searching any selective page so why bother to give this control to the user.

Reply:
Is it the goal for this SDK to facilitate access to the underlying Nimble REST API, or to abstract it for particular use cases?

Reply:
Pagination is a SDK feature.

Reply:
@rkumpf , we want the SDK to make it easy to access the underlying storage resources.  Ideally, if the user simply requests all objects, the SDK ought to internally page and report all objects.  If the user specifies paging details, we should apply that to our API request and report paging details in the response.

Reply:
Thanks, @rgcostea.  Interesting point about internally looping to fulfill the request page parameters if the array is unwilling to do so in a single request.  It does not appear to be part of the current design.  The same principle could apply when a caller requests rows 10 to 10,000,000, for example.  Perhaps a caller could explicitly request this convenience, to indicate acceptance of the response time implications.

This discussion concerns the page-level interface exposed by the SDK.  My aim is to avoid what I perceive as confusing redundancy in the page parameters while enabling a user to access the capabilities of the underlying API.  I.e., make it simple to loop through all the pages while still providing control to request rows 10 to 30, for example.

To achieve this, I'm suggesting: 1) Separate the retrieved page info from the requested page parameters (instead of modifying the input parameters with output data), 2) Move the PageSize attribute to a convenience function which can translate it to StartRow and EndRow, and 3) Give the caller control over subsequent pages rather than making assumptions inside the SDK.

Reply:
> It does not appear to be part of the current design
It needs to be in the final product.  We do this in our python and java SDKs.

I feel all 3 of your suggestions are sound.  Keeping the SDK as simple as possible for the end user should be one of our overarching goals.

Reply:
How could implementing this be a simplified approach :
```
arg.Page = new(pagination.Page)
var retrievedPage pagination.Page

for hasNextPage := true; hasNextPage; hasNextPage = retrievedPage.HasNextPage() {
	arg.Page.NextPage(retrievedPage, 2)
	volumes, retrievedPage, err := suite.volumeService.GetVolumes(arg)
}
```
User has to constantly feed the page no which is un-necessary, are you suggesting to have a int variable and increment it inside the for loop to set the page no all the time ? This is not at all intuitive to user.  

Here is an example of C++ SDK and ours implementation similar to that 
` // "name"=="0" && "size">=1024  sortBy:size
    gp = GetParamsSmtPtr(new GetParams(sf3, sk), getParamsDeleter);
    PaginationParamsSmtPtr pgp = gp->getPaginationParams();
    pgp->setBatchSize(2);
    pgp->setPaginationStartIndex(3);
    gp->setPaginationParams(pgp);
    do {
        volList = volumeOS->getObjectList(gp);
        startRow = pgp->getPaginationStartIndex();
        endRow = pgp->getPaginationEndIndex();
        totalRows = pgp->getTotalRowCount();
        cout << "startRow: " << startRow << "   endRow: " << endRow;
        cout << "   totalRows: " << totalRows << "   batchSize: " << pgp->getBatchSize() << endl;
        //ensure that the number of rows returned never exceeds the specfied batchSize.
        assert((pgp->getPaginationEndIndex() - pgp->getPaginationStartIndex()) <= pgp->getBatchSize());
    } while(pgp->containsMore());`

I could have written the same unit test code like this with existing pagination package 
`
gp := new(param.GetParams)

pgp :=new(pagination.Page)
pgp.SetPageSize(2)

gp.Page = pgp
volumes, err := suite.volumeService.GetVolumes(gp)
for  pgp.ContainsMore() {
	volumes, err := suite.volumeService.GetVolumes(gp)
}
`

Reply:
The integer literal in the flow I suggested is a page size, not a page number.  A caller is not expected to track page numbers.

My principal aim is to clarify the interface.  For example, it is not obvious that this code would fetch rows 5 to 14:

    gp := new(param.GetParams)
    pgp :=new(pagination.Page)
    pgp.SetStartRow(5)
    pgp.SetEndRow(12)
    pgp.SetPageSize(10)
    gp.Page = pgp
    volumes, err := suite.volumeService.GetVolumes(gp)

And who would guess that this code would fetch rows 12 to 19?

    gp := new(param.GetParams)
    pgp :=new(pagination.Page)
    pgp.SetStartRow(5)
    pgp.SetEndRow(12)
    pgp.SetPageSize(10)
    x := 20
    pgp.totalRows = &x
    gp.Page = pgp
    volumes, err := suite.volumeService.GetVolumes(gp)

This one will panic:

    gp := new(param.GetParams)
    pgp :=new(pagination.Page)
    x := 20
    pgp.totalRows = &x
    gp.Page = pgp
    volumes, err := suite.volumeService.GetVolumes(gp)

One could argue "don't do that".  But in a general purpose SDK, and in open source no less, it is expected that the behavior is well defined.  Rather than document potentially bizarre behavior or limitations, it seems straightforward to remove the uncertainty from the interface.

I suggested a possible design to address the clarity concerns.  I am certainly open to alternatives.

Reply:
The user input validation logic is not built yet since this is something that should be done for all the interfacing functions /object. so kept it open for now . That will clarify all of these to sdk user .   
Here in case of pagination, either you set PageSize and StartRow or you set StartRow and EndRow that user input validation will be implemented before end product release. Currently it is done . 
The rest of the combination will throw error like PageSize , StartRow and EndRow. Only the above two listed combinations are allowed. 
If you notice I have provided only user settable options for PageSize , Start Row and Endrow . The TotalRow is not even allowed to be set by the user.

Reply:
This caller code construction was posited:

	for  pgp.ContainsMore() {
		volumes, err := suite.volumeService.GetVolumes(gp)
	}

I feel the interface does not make it clear that the pgp values are modified by the GetVolumes() call, both before they are used in a REST API call and after a response is received.  The conflation of input parameters and output results seems unexpected and suboptimal.

Reply:
@rkumpf, as I see you have only issue with the conflation of input/output parameters. As I posted earlier, this is similar to what we do in CPP SDK and maybe in others. I posted the example as well. How do you propose to change it ?

Reply:
I expressed these concerns:

* Redundant/overlapping attributes in the Page model.
* Input parameters modified with "next page" logic that is hidden from the caller and outside their control.
* Input parameters unexpectedly modified with output values.

My suggestions to address these concerns are:

1. In ListFromParams (and in the Get functions that call it), add a third return value with the retrieved page values.  Do not modify the input parameters.
1. Remove the PageSize attribute from the Page structure.
1. Add a convenience function that sets appropriate StartRow and EndRow values in a Page structure based on the page values from a prior Get call (if applicable) and a given page size.  The NextPage function in an earlier comment is a possible implementation.
1. In ListFromParams, use the StartRow and EndRow values provided as input, without modification.

Reply:
#1 )I am not convinced with your suggestion, earlier you said there is a conflation of input/output and as per your suggestion, we are endup doing the same thing in a different way. Here is your example
```
func (page *Page) NextPage(retrievedPage Page, pageSize int) {
	if retrievedPage == nil || retrievedPage.EndRow == nil {
		page.SetStartRow(0)
	} else {
		page.SetStartRow(*retrievedPage.EndRow)
	}
	page.SetEndRow(*page.StartRow + pageSize)
}
```
Aren't we modifying the in StartRow and EndRow here internally? we are modifying the input and output here . So conflation of i/o remains. 

#2 , The suggested change will have an additional return type retrivedPage
```
volumes, retrievedPage, err := suite.volumeService.GetVolumes(arg)
```

The paginartion is optional feature and user may not wish to set this , so this third argument is always going to be redundant and will have to pass nil all the time un-necessarily. In addition to it , this will have impact on all the services /object and objectset . I don't see much value here .

#3)

User has to make additional call to the fetch the nextPage() , here in your code example 
```
arg.Page = new(pagination.Page)
var retrievedPage pagination.Page
for hasNextPage := true; hasNextPage; hasNextPage = retrievedPage.HasNextPage() {
	arg.Page.NextPage(retrievedPage, 2)
	volumes, retrievedPage, err := suite.volumeService.GetVolumes(arg)
}
```
I don't think this is required as the approach i implemented is neat and easily consumable .
Here is an example 
```
arg.Page = new(pagination.Page)
arg.Page.SetPageSize(2)
volumes, err := suite.volumeService.GetVolumes(arg)
for arg.Page.ContainsMore() {
        volumes, err := suite.volumeService.GetVolumes(arg)
}
```

@rgcostea , I leave it to you to make the call.

Reply:
Is the server obligated to return all the rows requested?  I wasn't able to discern it from the Nimble REST API documentation.  I've seen in other APIs where the server has discretion over how many rows to return.  For example, if a client requests a million rows, and the server determines it to be excessive, it could elect to return only 100.

Regarding the NextPage function, I wouldn't characterize these updates to the page attributes as "internal", because this function is part of the model within the caller's control.  I see this additional client call as a benefit, while you see it as extra complexity.  Oh well.  :)

Reply:
> Is the server obligated to return all the rows requested? I wasn't able to discern it from the Nimble REST API documentation. I've seen in other APIs where the server has discretion over how many rows to return. For example, if a client requests a million rows, and the server determines it to be excessive, it could elect to return only 100.
> 
> Regarding the NextPage function, I wouldn't characterize these updates to the page attributes as "internal", because this function is part of the model within the caller's control. I see this additional client call as a benefit, while you see it as extra complexity. Oh well. :)

As far as I know the Nimble REST API behavior, it returns all.

Reply:
If we can confirm the server always returns all requested rows, that simplifies things.  The startRow and endRow values in the response become irrelevant, so we don't have to worry about returning them to the caller.  In that case, I suggest:

1. Remove the PageSize attribute from the Page structure.
2. Implement the SetPageSize function to set StartRow=0 and EndRow=pageSize.
3. In ListFromParams, use the StartRow and EndRow values provided as input, without modification.
4. In ListFromParams, set the TotalRows value in the Page parameter according to the response from the array.  Leave the StartRow and EndRow values unchanged.
5. Change the ContainsMore() function to a NextPage() function that has a dual role of updating the StartRow and EndRow values for the next page and returning a boolean indicating whether there actually is a next page.  For example:

```
func (page *Page) NextPage() boolean {
	if page.StartRow == nil || page.EndRow == nil || page.TotalRows == nil ||
		*page.EndRow >= *page.TotalRows || *page.StartRow >= *page.EndRow {
		return false
	}
	pageSize := *page.EndRow - *page.StartRow
	page.startRow = page.EndRow
	page.SetEndRow(*page.StartRow + pageSize)
	return true
}
```

A caller who chooses to use paging could then use the same logic as the existing implementation, with just one method name change:

```
arg.Page = new(pagination.Page)
arg.Page.SetPageSize(2)
volumes, err := suite.volumeService.GetVolumes(arg)
for arg.Page.NextPage() {
	volumes, err := suite.volumeService.GetVolumes(arg)
}
```

These minor changes would address all three of the concerns I raised earlier:

* Redundant/overlapping attributes in the Page model.
* Input parameters modified with "next page" logic that is hidden inside the Get function and outside the caller's control.
* Input parameters unexpectedly modified with output values.

Again, this assumes the array honors requests for any number of rows.

Reply:
This is exactly what I have implemented, it just that startRow & endRow computation logic has moved from group_mgmt_client.go to pagination.go package. Also, the page field is removed from the struct . 

I accept this change as it makes more sense to have this logic in pagination.go file.